lnrpc/walletrpc/walletkit_server.go

```diff
@@ -819,3 +840,226 @@ func (w *WalletKit) LabelTransaction(ctx context.Context,
+		packet, err = psbt.NewFromRawBytes(r, false)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse PSBT: %v", err)
+		}
+
+	// The template is specified as a RPC message. We need to create a new
+	// PSBT and copy the RPC information over.
+	case req.GetRaw() != nil:
+		tpl := req.GetRaw()
+		if len(tpl.Outputs) == 0 {
+			return nil, fmt.Errorf("no outputs specified")
+		}
+
+		txOut := make([]*wire.TxOut, len(tpl.Outputs))
+		for idx, out := range tpl.Outputs {
+			txOut[idx] = &wire.TxOut{
+				Value:    out.Value,
+				PkScript: out.PkScript,
+			}
+		}
+
+		txIn := make([]*wire.OutPoint, len(tpl.Utxos))
+		for idx, in := range tpl.Utxos {
+			op, err := unmarshallOutPoint(in)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing "+
+					"outpoint: %v", err)
+			}
+			txIn[idx] = op
+		}
+
+		packet, err = psbt.New(txIn, txOut, 2, 0, nil)
+		if err != nil {
+			return nil, fmt.Errorf("could not create PSBT: %v", err)
+		}
+
+	default:
+		return nil, fmt.Errorf("transaction template missing, need " +
+			"to specify either PSBT or raw TX template")
+	}
+
+	// In case the user did specify inputs, we need to make sure they are
+	// known to us, still unspent and not yet locked.
+	if len(packet.UnsignedTx.TxIn) > 0 {
+		// Get a list of all unspent witness outputs. We'll acquire the
+		// global coin selection lock to ensure there aren't any other
+		// concurrent processes attempting to lock any UTXOs which may
+		// be shown available to us.
+		var utxos []*lnwallet.Utxo
+		err = w.cfg.CoinSelectionLocker.WithCoinSelectLock(func() error {
+			utxos, err = w.cfg.Wallet.ListUnspentWitness(
+				defaultMinConf, defaultMaxConf,
+			)
+			return err
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// Validate all inputs against our known list of UTXOs now.
+		err = addInputUTXOInfo(packet, utxos)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Determine the desired transaction fee.
+	switch {
+	// Estimate the fee by the target number of blocks to confirmation.
+	case req.GetTargetConf() != 0:
+		targetConf := req.GetTargetConf()
+		if targetConf < 2 {
+			return nil, fmt.Errorf("confirmation target must be " +
+				"greater than 1")
+		}
+
+		feeSatPerKW, err = w.cfg.FeeEstimator.EstimateFeePerKW(
+			targetConf,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("could not estimate fee: %v",
+				err)
+		}
+
+	// Convert the fee to sat/kW from the specified sat/vByte.
+	case req.GetSatPerVbyte() != 0:
+		feeSatPerKW = chainfee.SatPerKVByte(
+			req.GetSatPerVbyte() * 1000,
+		).FeePerKWeight()
+
+	default:
+		return nil, fmt.Errorf("fee definition missing, need to " +
+			"specify either target_conf or set_per_vbyte")
+	}
+
+	// We made sure the input from the user is as sane as possible. We can
+	// now ask the wallet to fund the TX. This will not yet lock any coins
+	// but might still change the wallet DB by generating a new change
+	// address.
+	err = w.cfg.Wallet.FundPsbt(packet, feeSatPerKW)
```

Comment:
This should be wrapped in the main coin selection context manager (the mutex).

Reply:
You mean `WithCoinSelectLock`, right? Wrapped all wallet operations in a closure.